media-source/webm/test-v-no-keyframe-cluster-manifest.json

```diff
@@ -0,0 +1,4 @@
+{
+    "url": "webm/webm-no-keyframe-cluster.webm",
```

Comment:
Inspecting the binary webm file using hachoir-urwid:
It looks like it contains 2 tracks (track 1 = video, track 2 = audio).
Cluster[0]:
SimpleBlock[0]: track 1 (video), keyframe bit is SET
SimpleBlock[1 through 3, and 5 through 6]: track 2 (audio), keyframe bit is NOT SET
SimpleBlock[4]: track 1 (video), keyframe bit is NOT SET

Cluster[1] has 13 SimpleBlocks, following same pattern as Cluster[0]: first video frame is a keyframe, remainder of video and all of audio are not keyframes.

There are 53 total clusters in the webm file -- I didn't go through all of them.
I then ran the file through a debug-build of tip-of-tree Chromium with logging to see what the Chromium MSE parsers see.

Items in the binary that I believe need fixing:
1) All audio frames should be marked keyframe. Chromium is lenient and updates those in the parser to be keyframe. Other implementations **aren't required to be so lenient**. (You can see examples of Chromium noticing this from the player's log in chrome://media-internals having entries like: "Bytestream with audio frame PTS 0us and DTS 0us indicated the frame is not a random access point (key frame). All audio frames are expected to be key frames.")

2) These clusters' first video track frame (at least for the first couple clusters) is a keyframe. This sounds like the reverse of the scenario you intend to test, unless I'm mistaken.

3) All of the media is contained in SimpleBlocks within Clusters. Notably, having a SimpleBlock at the end of a Cluster in a muxed multi-track stream leads to inability of implementations to understand the intended duration of that block at the end of the cluster (earlier block durations in the cluster for that track can be inferred by timestamp delta; not so with the last block). This can result in the MSE coded frame processing algorithm's discontinuity detection logic kicking in, forcing all tracks to drop nonkeyframes until their next keyframe is received. So, appending this media in Chrome M64 leads to 5 disjoint buffered ranges (due to video nonkeyframes being dropped after a perceived discontinuity due to under-guessed audio block durations (5ms, instead of 23ms or 24ms) from the end of one cluster to the beginning of the next). In tip-of-tree/Canary M66, there's logic that helps a little: those nonkeyframes are still dropped, but the gaps are coalesced better. Regardless, the guidance for muxers of WebM is to use BlockGroup (which encode explicit BlockDuration) for at least the last block for each track in a cluster. That avoids requiring implementations to guess those durations (which can result in unexpected gaps/stalls/dropped nonkeyframes).

Reply:
Item 3 can be alleviated if you need by
a) using DefaultDuration to be close to the actual block durations, or have all the blocks have very similar durations. Unfortunately, the first audio block in the first cluster was 5ms, unlike the rest that I inspected (23-24ms), and
b) using just one media track per SourceBuffer (e.g., just video for your case, since audio can't have nonkeyframes...)